### ISSUE TITLE
Introduce "ignoreForSingleProperty" value for prefer-destructuring

### ISSUE THREAD

**Issue by rugk** [eslint/eslint#10250]
**What rule do you want to change?** [`prefer-destructuring`](https://eslint.org/docs/rules/prefer-destructuring)

**Does this change cause the rule to produce more or fewer warnings?** if configured, fewer; otherwise no change

**How will the change be implemented? (New option, new default behavior, etc.)?** new option

**Please provide some example code that this change will affect:**

Web Extension here:
```js
browser.storage.sync.get("popupIconColored").then((res) => {
    const popupIconColored = res.popupIconColored;

    // …
});
```
([real-life example](https://github.com/rugk/offline-qr-code/blob/a15adbc1c8e93327579128bbc54bef61a4d4651f/src/background.js#L49-L58))

**What does the rule currently do for this code?**
When `prefer-destructuring` is enabled, it complains that I should use destructuring here. So I'd have to use this:
```js
const { popupIconColored } = res;
``` 

(not tested, it also has been mentioned [I may have to put brackets around that even more…](https://github.com/eslint/eslint/issues/10067#issuecomment-371051079))

**What will the rule do after it's changed?**

It should not complain about the previous code example.

I would like a config option named "smart" or similar, so that the rule only catches cases, where destructuring "makes sense" (hereby I mean that I access multiple values of an array/object of it directly) and not, when I only access one value and want a short variable name for that.

[This example e.g. makes sense](https://github.com/airbnb/javascript#destructuring) and should still be catched with the `smart` setting:
```js
function getFullName(user) {
  const firstName = user.firstName;
  const lastName = user.lastName;

  return `${firstName} ${lastName}`;
}
```

Because here, I assign/get two variables from the object. However, when I only want to get one variable, destructuring just _complicates the code_ IMHO.

**In short:** Add an option, which only triggers when I get multiple values from an array/object and not only a single one.

**Comment by ljharb** [comment]
I disagree very much with the subjective “makes sense” there; destructuring is helpful even for one value, because avoiding typing the name twice reduces bugs.

**Comment by rugk** [comment]
Ah, okay, I see. That's an argument. (I also fixed my code above, as I now understood that destructuring :wink:). Anyway, yes, of course what you prefer is subjective – that's the whole reasons for such options.

So the smart option would be for people, who do not care about the "type twice" aspect of that, but rather about not having multiple lines of code for assigning properties an object/array to a variable. And BTW, that "makes code complicated" was [not only my idea](https://github.com/eslint/eslint/issues/10067#issuecomment-377775794).

**Comment by ljharb** [comment]
The issue you are linking to has nothing to do with that - it’s solely about destructuring for assignment exprsssions (no var/let/const), which can already be controlled with an option.

**Comment by rugk** [comment]
Ah, that's the `enforceForRenamedProperties`. Okay, good to know, thanks.

**Comment by ljharb** [comment]
No, that's the "AssignmentExpression" option.

`enforceForRenamedProperties` is for controlling whether you allow `const foo = obj.bar` or whether you require it to be `const { foo: bar } = obj;`.

**Comment by rugk** [comment]
> // Now what if your statements are split? Should this warn?

Probably, yes, as JS does hoisting (or how it is called). But there is another ESLint rule about hosting/vars on top anyway, so I am not sure…
Or ait… hoisting is not done for constants, is not it?

**Comment by ljharb** [comment]
Only the declaration is hoisted (let and const have a TDZ); the initialization never is.

**Comment by platinumazure** [comment]
I'm a soft :-1: for this due to how complex the implementation would likely have to be to make this work in a way that "makes sense" to people likely to use it.

@rugk If this proposal does not get accepted by the ESLint team, you could probably work around this and get a rule that does what you want by designing a custom rule which could run this rule and then filter out reports that end up only applying for a variable (or object/property) once in the same scope. You could use something like [eslint-rule-composer](https://github.com/not-an-aardvark/eslint-rule-composer) to leverage the core `prefer-destructuring` and then simply filter out messages that shouldn't actually count (i.e., they are 1-level destructuring and no other lint message references the same object). I think that would actually be a lot simpler to implement at this point than a change to the core rule.

**Comment by rugk** [comment]
Thanks for that information. I think it is not really worth it for such a simple rule (I then just rather don't use that rule.), but thanks for the tip.
